amcl/src/amcl_node.cpp

```diff
@@ -1602,13 +1623,20 @@ AmclNode::laserReceived(const sensor_msgs::LaserScanConstPtr& laser_scan)
         odom_amcl_diff_z_pub_.publish(dz_msg);
         diff_count_pub_.publish(diff_count_msg);
 
-        if (diff_count_ >= odom_amcl_diff_count_thre_) diff_count_ = 0;
+        if (diff_count_ >= odom_amcl_diff_count_thre_)
+        {
+          diff_count_ = 0;
+          pure_amcl_delta_sum_ = pf_vector_zero();
+          pure_odom_delta_sum_ = pf_vector_zero();
+        }
```

Comment:
Instead of resetting, I think it would be more appropriate to use a ring buffer to discard the oldest data and move on to the next one as a monitoring method.

----

_リセットするのではなく、リングバッファのような形で、最古のデータを捨てて、次に進む方が監視の仕方としてあっている気がする。_

Reply:
そうですね、それはこちらも思いました。
打ち合わせで話していた、たまたま合致したタイミングでそれまでのズレを全て受け入れてしまう事になってしまうのではないかな？と。

Reply:
ちょっと考えてみたのですが、リングバッファのような形は今回のロジックに適さないのではないかと思います。

たまたま合致したタイミングでそれまでのズレを全て受け入れてしまう事の対策として、過去のズレの累積と、(累積分のステップ数) x (1ステップ分の閾値) を比較するように変更しています。

リングバッファで貯めたpure_odom_delta と pure_amcl_delta の差分は、odom correction によって、0になるので、pf_vector_zero() を実行したことと同じになります。

今回の指摘とは異なる改善案ですが、3ステップ分の累積が5ステップ分の閾値さえも超えてしまっている、みたいなケースについて考えました。現状のロジックでは5ステップまで待たないとodom correction が実行されないのですが、この場合は5ステップ待つまでもなく3ステップ目の段階でodom correction を実行するべきだと思ったので、実装に組み込もうと思います。

Reply:
odom crectionしているときにdiff_count_を0リセットできているので、同意です。

個人的にわかりやすくするとしたら、ズレていない場合にdiff_count_を0リセットするように書き換えることでしょうか。( このリセット部分は消す。 )
そして、いざズレたときにdiff_count_をインクリメントする。

----

I agree, as I am able to reset diff_count_ to 0 when I am odom crectioning.

If I were to make it personally clearer, I would rewrite it so that diff_count_ is reset to 0 when it is not misaligned. ( I would erase this reset part. )
And then when it comes time to misalign, increment diff_count_.

Reply:
> 現状のロジックでは5ステップまで待たないとodom correction が実行されないのですが、この場合は5ステップ待つまでもなく3ステップ目の段階で odom correction を実行するべきだと思ったので、実装に組み込もうと思います。

immidiate_odom_updateと名付けたのですが、
実装してみると、immidiate_odom_update が連続して実行されすぎてしまうという問題がありました。odom correction 後は一定の期間(少なくとも5ステップ)空いている方が精度よかったです。
論理的な説明はまだ用意できてないのですが、rosbagだけアップしておきます。
[Screencast from 01-22-2024 07:58:42 PM.webm](https://github.com/LexxPluss/navigation/assets/73733804/ec3e9ff4-05c7-417b-bff4-fe2e0d8a4287)

Reply:
↑rqt_multiplot の 右下のdiff_countが6になるとimmidiate_odom_correction がかかったことを意味する。

Reply:
> immidiate_odom_updateと名付けた

immediate